test/controllers/tasks_controller_test.rb

```diff
@@ -1,126 +1,164 @@
+      get edit_task_path(task.id)
+      must_respond_with :success 
     end
 
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+      get edit_task_path(500)
+      must_redirect_to root_path
     end
   end
 
-  # Uncomment and complete these tests for Wave 3
-  describe "update" do
-    # Note:  If there was a way to fail to save the changes to a task, that would be a great
-    #        thing to test.
+  describe "update action" do
     it "can update an existing task" do
-      # Your code here
+      existing_task = Task.create(name:"Feed Kaya", description: "She's a fatty. Feed half a cup a meal and lots of snacks.")
+      updated_task_data = {
+        task: {
+          name: "Feed Kaya",
+          description: "She's a fatty. Feed half a cup a meal and lots of snacks -- don't forget 1 dental a day.",
+          completed: nil
+        }
+      }
+
+      patch task_path(existing_task.id), params: updated_task_data
+
+      updated_task = Task.find_by(id: existing_task.id)
+      expect(updated_task.name).must_equal updated_task_data[:task][:name]
+      expect(updated_task.description).must_equal updated_task_data[:task][:description]
+      expect(updated_task.completed).must_be_nil updated_task_data[:task][:completed]
+
+      must_respond_with :redirect
+      must_redirect_to task_path(updated_task.id)
     end
 
     it "will redirect to the root page if given an invalid id" do
-      # Your code here
+      patch task_path(500)
+      must_redirect_to root_path 
     end
   end
 
-  # Complete these tests for Wave 4
-  describe "destroy" do
-    # Your tests go here
+  # Wave 4
+  describe "destroy action" do
+    it "successfully deletes an existing Task and then redirects to home page" do
+      Task.create(name: "Cuddle with Kaya", description: "She hates cuddles but do it anyways hehehehehe", completed: nil)
+      existing_task_id = Task.find_by(name: "Cuddle with Kaya").id
+
+      expect {
+        delete task_path(existing_task_id)
+      }.must_differ "Task.count", -1
+
+      must_redirect_to root_path
+    end
+
+    it "redirects to tasks index page and doesn't delete any tasks if no tasks exist" do
+      Task.destroy_all
+      invalid_id = 1
+
+      expect {
+        delete task_path(invalid_id)
+      }.must_differ "Task.count", 0
+
+      must_redirect_to root_path
+    end
+
+    it "redirects to tasks index page if selected task has already been deleted and doesn't delete any other tasks" do
+      my_task = Task.create(name: "Cuddle with Kaya", description: "She hates cuddles but do it anyways hehehehehe", completed: nil)
+      
+      Task.destroy_all
+
+      expect {
+        delete task_path(my_task.id)
+      }.must_differ "Task.count", 0
+
+      must_redirect_to root_path
+    end
 
   end
 
-  # Complete for Wave 4
-  describe "toggle_complete" do
-    # Your tests go here
+  describe "completed action" do
```

Comment:
An edge case to test for would be trying to use the completed action for a task that didn't exist